man/liebert-esp2.8
.TH liebert-esp2 8 "Fri Apr 09 2010" "" "Network UPS Tools (NUT)"
.SH NAME  
liebert-esp2 \- Driver for Liebert UPS, using the ESP2 serial protocol
.SH NOTE
This man page only documents the hardware\(hyspecific features of the
liebert-esp2 driver.  For information about the core driver, see  
\fBnutupsdrv\fR(8).

.SH SUPPORTED HARDWARE
This is an experimental driver.  You have been warned.

.SH EXTRA ARGUMENTS
This driver supports the following optional settings in \fBups.conf\fR(5):

.IP "baudrate=\fInum\fR"
Set the speed of the serial connection - 1200, 2400 (default), 4800, 9600 or 19200.

.SH AUTHOR
Richard Gregory <R.Gregory at liverpool.ac.uk>, Arjen de Korte <adkorte\(hyguest at alioth.debian.org>

.SH SEE ALSO

.SS The core driver:
\fBnutupsdrv\fR(8)

.SS Internet resources:
The NUT (Network UPS Tools) home page: http://www.networkupstools.org/


NEWS
If you're upgrading from an earlier version, see the UPGRADING file.

For a complete list of changes, please refer to the ChangeLog file.

---------------------------------------------------------------------------
Release notes for NUT 2.4.3 - what's new since 2.4.2:

 - this is a bugfix release that only solves the regression on IPv6 activation.

---------------------------------------------------------------------------
Release notes for NUT 2.4.2 - what's new since 2.4.1:

 - the general USB support has been vastly improved, including many bug
   fixes, better OS support, new features and devices.

 - NUT now talks to Solar Controller Devices with the new ivtscd driver.

 - the snmp-ups driver supports more PDU, with a smaller disk footprint.

 - apcsmart supports more older SmartUPS and Matrix units.

 - the bestfortress driver is resurrected.

 - the virtual driver has been renamed to 'clone'.

 - the netxml-ups driver has received some care.

 - various debugging and development improvements have been done, around
   driver output; dummy-ups with more interaction and scripting and the
   device-recorder.sh script.

 - the build system has received many bugfixes and improvements.

 - the UPower (previously known as DeviceKit-power) rules file is now
   generated by NUT.

 - support for new devices: Apollo 1000A and 1000F; various Baytech RPC; old
   Best Power Fortress; Cyber Power Systems PR3000E, CP 1500C and OR2200LCDRM2U;
   all the new Dell UPS range (serial, USB and network); Eaton E Series NV and
   DX UPS, and Powerware 9130; older HP T500 and T750, newer T750 INTL (USB) and
   R1500 G2 (serial); Inform Informer Compact 1000VA; many serial and USB
   devices from Ippon, like Back Comfo Pro, Smart Power Pro and Smart Winner;
   IVT SCD series; Liebert GXT2-3000RT230 and PowerSure PSA; Mustek PowerMust
   424 / 636 / 848 USB; all new PowerCOM USB devices with HID PDC interface;
   Tripp-Lite INTERNETOFFICE700, SMART700USB and ECO550UPS; UPSonic DS-800
   (USB).

---------------------------------------------------------------------------
Release notes for NUT 2.4.1 - what's new since 2.4.0:

 - the microdowell driver has appeared to support various MicroDowell Enterprise
   units (see the "new devices" list below).

 - support for new devices: MicroDowell Enterprise B8, B10, N8, N11, N15, N20,
   N22, N30, N40, N50, N60 and HiBox ST.

 - NUT-Monitor now better handles the ups.status field, and has switched to
   version 1.1.

 - the situation of the build toolchain has been fixed, with regard to the
   "make clean" target and the wrongly removed generated USB files. This brokes
   further configure call. 

---------------------------------------------------------------------------
Release notes for NUT 2.4.0 - what's new since 2.2.2:

 - preliminary support for Power Distribution Units (PDUs): NUT now support
   PDUs, either natively (ie using NUT snmp-ups driver), or through a binding to
   the Powerman daemon. The list of supported PDUs is already quite long,
   including: Eaton ePDUs (Managed and Monitored), some Aphel models, some
   Raritan PDUs, and the whole list of Powerman supported devices:
   http://powerman.sourceforge.net/supported.html
 
 - support for new devices: the various PDUs cited above, Chloride Desk Power
   650, Cyber Power Systems Value 400E/600E/800E (USB models), Delta GES602N,
   Digitus DN-170020, the whole Eaton ranges (mostly composed of MGE Office
   Protection Systems and Powerware units) including BladeUPS, Forza Power
   Technologies SL-1001, HP PowerTrust 2997A, HP R/T 2200 G2, Infosec XP 1000
   and XP 500, Ippon Back Power Pro (serial and USB), Kebo 1200D/D Series,
   Liebert PowerSure Personal XT, MGE Office Protection Systems Protection
   Station, Neus 400va and 600va, Phasak 400VA and 600VA, Plexus 500VA, Powercom
   Black Knight PRO / King PRO and Imperial, PowerKinetics BlackOut Buster,
   Sweex 1000 USB, UNITEK Alpha 500, WinPower CPM-800.

 - NUT now embeds Python client support through the PyNUTClient module and the
   NUT-Monitor application. Both are from David Goncalves, and are still
   available from http://www.lestat.st.
   For more information, refer to scripts/python/README.

 - the dummy-ups driver now support a "repeater" mode. This allows it to act as
   a NUT client, and to forward data. This can be useful for supervision and
   load sharing purposes.

 - tcp-wrappers support has been added to the upsd server, to grant users access
   by source IP for commands that require to be logged into the server. This
   replaces the previous internal implementation (ACL in upsd.conf).

 - the nut.conf file has been introduced to standardize startup configuration
   across the various systems.

 - NUT now ships a bash completion function for 'upsc' command
   (scripts/misc/nut.bash_completion). Simply copy it to /etc/bash_completion.d
 
 - many internal changes to improve maintenability, while lowering the
   maintenance cost (thus allowing developers to focus on what matters: the
   code!). Examples of this are:
   - the USB information automatic extraction to generate the various USB helper
     files,
   - the upsdrv_info_t structure to track more driver information, and remove
     the need for the upsdrv_banner() function
   - common USB code refactoring, as it is done for the serial functions.

 - tons of bugfixes, cleanup and improvements to make NUT stronger than ever!

---------------------------------------------------------------------------
Release notes for NUT 2.2.2 - what's new since 2.2.1:

 - support for new devices: APC BACK-UPS XS LCD, Atlantis Land,
   Mustek Powermust Office 650, Oneac XAU models, Powerware PW5115 and
   PW9120 (USB), Nitram Elite 2005
 
 - Integrated Power Management (NUT HAL integration) has reached a
   major milestone: it is now the most advanced UPS integration into
   Power Management layer known in existing OSs. It has received many

drivers/liebert-esp2.c
/* liebert-esp2.c - driver for Liebert UPS, using the ESP-II protocol
 *
 *  Copyright (C)
 *  2009	Richard Gregory <r.gregory liverpool ac uk>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */

#include "main.h"
#include "serial.h"
#include "timehead.h"
#include "nut_stdint.h"

#define DRIVER_NAME	"Liebert ESP-II serial UPS driver"
#define DRIVER_VERSION	"0.02"

static int instcmd(const char *cmdname, const char *extra);
static int setvar(const char *varname, const char *val);

/* driver description structure */
upsdrv_info_t upsdrv_info = {
	DRIVER_NAME,
	DRIVER_VERSION,
	"Richard Gregory <r.gregory liv ac uk>",
	DRV_EXPERIMENTAL,
	{ NULL }
};

static const unsigned char
	/* Bit field information provided by Spiros Ioannou */
	/* Ordered on MSB to LSB. Shown as DESCRIPTION (bit number), starting at 0. */
	cmd_bitfield1[]		= { 1,148,2,1,1,153 },	/* ON_BATTERY(8), INPUT_OVERVOLTAGE(7), BATTERY_TEST_STATE(6), OVERTEMP_WARNING(5), INRUSH_LIMIT_ON(4), UTILITY_STATE(3), ON_INVERTER(2), DC_DC_CONVERTER_STATE(1), PFC_ON(0) */
	cmd_bitfield2[]		= { 1,148,2,1,2,154 },	/* BUCK_ON (9), DIAG_LINK_SET(7), BOOST_ON(6), REPLACE_BATTERY(5), BATTERY_LIFE_ENHANCER_ON(4), BATTERY_CHARGED (1), ON_BYPASS (0) */
	cmd_bitfield3[]		= { 1,148,2,1,3,155 },	/* CHECK_AIR_FILTER (10), BAD_BYPASS_PWR (8), OUTPUT_OVERVOLTAGE (7), OUTPUT_UNDERVOLTAGE (6), LOW_BATTERY (5), CHARGER_FAIL (3), SHUTDOWN_PENDING (2), BAD_INPUT_FREQ (1), UPS_OVERLOAD (0) */
	cmd_bitfield7[]		= { 1,148,2,1,7,159 },	/* AMBIENT_OVERTEMP (2) */
	cmd_battestres[]	= { 1,148,2,1,12,164 },	/* BATTERY_TEST_RESULT */
	cmd_selftestres[]	= { 1,148,2,1,13,165 };	/* SELF_TEST_RESULT */

static char cksum(const char *buf, const size_t len)
 {
	char	sum = 0;
	size_t	i;

	for (i = 0; i < len; i++) {
		sum += buf[i];
	}

	return sum;
}

static int do_command(const unsigned char *command, char *reply)
{
	int	ret;

	ret = ser_send_buf(upsfd, command, 6);
	if (ret < 0) {
		upsdebug_with_errno(2, "send");
		return -1;
	} else if (ret < 6) {
		upsdebug_hex(2, "send: truncated", command, ret);
		return -1;
	}

	upsdebug_hex(2, "send", command, ret);

	ret = ser_get_buf(upsfd, reply, 8, 1, 0);
	if (ret < 0) {
		upsdebug_with_errno(2, "read");
		return -1;
	} else if (ret < 6) {
		upsdebug_hex(2, "read: truncated", reply, ret);
		return -1;
	} else if (reply[7] != cksum(reply, 7)) {
		upsdebug_hex(2, "read: checksum error", reply, ret);
		return -1;
	}

	upsdebug_hex(2, "read", reply, ret);
	return ret;
}

void upsdrv_initinfo(void)
{
	struct {
		const char	*var;
		const int	len;
	} vartab[] = {
		{ "ups.model", 15 },
		{ "ups.firmware", 8 },
		{ "ups.serial", 10 },
		{ "ups.mfr.date", 4 },
		{ NULL }
	};

	char	buf[LARGEBUF];
	int	i, index;

	dstate_setinfo("ups.mfr", "%s", "Liebert");

	for (i = 0; i < 37; i++) {
		char	command[6], reply[8];
		int	ret;

		snprintf(command, sizeof(command), "\x01\x88\x02\x01%c", i+4);
		command[5] = cksum(command, 5);

		ret = do_command((unsigned char *)command, reply);
		if (ret == 8) {
			buf[i<<1] = reply[6];
			buf[(i<<1)+1] = reply[5];
		} else {
			buf[i<<1] = '\0';
			buf[(i<<1)+1] = '\0';
		}
	}

	buf[i<<1] = 0;

	if (strlen(buf) == 0) {
		fatalx(EXIT_FAILURE, "ESP-II capable UPS not detected");
	}

	for (index = 0, i = 0; vartab[i].var; index += vartab[i].len, i++) {
		char	val[SMALLBUF];

		snprintf(val, sizeof(val), "%.*s", vartab[i].len, &buf[index]);

		dstate_setinfo(vartab[i].var, "%s", rtrim(val, ' '));
	}

	upsh.instcmd = instcmd;
	upsh.setvar = setvar;
}

void upsdrv_updateinfo(void)
{
	struct {
		const unsigned char	cmd[6];
		const char	*var;
		const char	*fmt;
		const double	mult;
	} vartab[] = {
		{ { 1,149,2,1,4,157 },	"battery.charge", "%.0f", 1.0 },
		{ { 1,149,2,1,1,154 },	"battery.runtime", "%.0f", 60 },
		{ { 1,149,2,1,2,155 },	"battery.voltage", "%.1f", 0.1 },
		{ { 1,149,2,1,3,156 },	"battery.current", "%.2f", 0.01 },
		{ { 1,149,2,1,12,165 },	"battery.temperature", "%.1f", 0.1 },
		{ { 1,161,2,1,13,178 },	"battery.voltage.nominal", "%.1f", 0.1 },
		{ { 1,149,2,1,7,160 },	"ups.load", "%.0f", 1.0 },
		{ { 1,149,2,1,6,159 },	"ups.power", "%.0f", 1.0 },
		{ { 1,161,2,1,8,173 },	"ups.power.nominal", "%.0f", 1.0 },
		{ { 1,149,2,1,5,158 },	"ups.realpower", "%.0f", 1.0 },
		{ { 1,149,2,1,14,167 },	"ups.temperature", "%.1f", 0.1 },
		{ { 1,144,2,1,1,149 },	"input.voltage", "%.1f", 0.1 },
		{ { 1,149,2,1,8,161 },	"input.frequency", "%.1f", 0.1 },
		{ { 1,149,2,1,10,163 },	"input.frequency.nominal", "%.1f", 0.1 },
		{ { 1,144,2,1,5,153 },	"input.bypass.voltage", "%.1f", 0.1 },
		{ { 1,144,2,1,3,151 },	"output.voltage", "%.1f", 0.1 },
		{ { 1,149,2,1,9,162 },	"output.frequency", "%.1f", 0.1 },
		{ { 1,144,2,1,4,152 },	"output.current", "%.1f", 0.1 },
		{ { 0 }, NULL, NULL, 0 }
	};

	const char	*val;
	char	reply[8];
	int	ret, i;

	for (i = 0; vartab[i].var; i++) {
		int16_t	val;

		ret = do_command(vartab[i].cmd, reply);
		if (ret < 8) {
			continue;
		}

		val = (unsigned char)reply[5];
		val <<= 8;
		val += (unsigned char)reply[6];

		dstate_setinfo(vartab[i].var, vartab[i].fmt, val * vartab[i].mult);
	}

	status_init();

	ret = do_command(cmd_bitfield1, reply);
	if (ret < 8) {
		upslogx(LOG_ERR, "Failed reading bitfield #1");
		dstate_datastale();
		return;
	}

	if (reply[5] & (1<<0)) {	/* ON_BATTERY */
		status_set("OB");
	} else {
		status_set("OL");
	}

	val = dstate_getinfo("battery.current");
	if (val) {
		if (atof(val) > 0.05) {
			status_set("CHRG");
		}
		if (atof(val) < -0.05) {
			status_set("DISCHRG");
		}
	}

	ret = do_command(cmd_bitfield2, reply);
	if (ret < 8) {
		upslogx(LOG_ERR, "Failed reading bitfield #2");
		dstate_datastale();
		return;
	}

	if (reply[6] & (1<<0)) {	/* ON_BYPASS */
		status_set("BYPASS");
	}

	if (reply[6] & (1<<5)) {	/* REPLACE_BATTERY */
		status_set("RB");
	}

	if (reply[6] & (1<<6)) {	/* BOOST_ON */
		status_set("BOOST");
	}

	if (reply[5] & (1<<1)) {	/* BUCK_ON */
		status_set("TRIM");
	}

	ret = do_command(cmd_bitfield3, reply);
	if (ret < 8) {
		upslogx(LOG_ERR, "Failed reading bitfield #3");
		dstate_datastale();
		return;
	}

	if (reply[6] & (1<<0) ) {	/* UPS_OVERLOAD */
		status_set("OVER");
	}

	if (reply[6] & (1<<5) ) {	/* LOW_BATTERY */
		status_set("LB");
	}

	status_commit();

	dstate_dataok();
}

void upsdrv_shutdown(void)
{
	/* replace with a proper shutdown function */
	fatalx(EXIT_FAILURE, "shutdown not supported");
}

static int instcmd(const char *cmdname, const char *extra)
{
/*
	if (!strcasecmp(cmdname, "test.battery.stop")) {
		ser_send_buf(upsfd, ...);
		return STAT_INSTCMD_HANDLED;
	}
 */
	upslogx(LOG_NOTICE, "instcmd: unknown command [%s]", cmdname);
	return STAT_INSTCMD_UNKNOWN;
}

static int setvar(const char *varname, const char *val)
{
/*
	if (!strcasecmp(varname, "ups.test.interval")) {

	ser_send_buf(upsfd, ...);
		return STAT_SET_HANDLED;
	}
 */
	upslogx(LOG_NOTICE, "setvar: unknown variable [%s]", varname);
	return STAT_SET_UNKNOWN;
}

void upsdrv_help(void)
{
}

/* list flags and values that you want to receive via -x */
void upsdrv_makevartable(void)
{
	addvar (VAR_VALUE, "baudrate", "serial line speed");
}

void upsdrv_initups(void)
{
	const char *val = getval("baudrate");
	speed_t baudrate = B2400;

	if (val) {
		switch (atoi(val))
		{
		case 1200:
			baudrate = B1200;
			break;
		case 2400:
			baudrate = B2400;
			break;
		case 4800:
			baudrate = B4800;
			break;
		case 9600:
			baudrate = B9600;
			break;
		case 19200:
			baudrate = B19200;
			break;
		default:
			fatalx(EXIT_FAILURE, "Baudrate [%s] unsupported", val);
		}
	}

	upsfd = ser_open(device_path);
	ser_set_speed(upsfd, device_path, baudrate);
}

void upsdrv_cleanup(void)
{
	ser_close(upsfd, device_path);
}
